src/server/pps/server/api_server.go

```diff
@@ -464,6 +467,248 @@ func (a *apiServer) RestartDatum(ctx context.Context, request *pps.RestartDatumR
 	return &types.Empty{}, nil
 }
 
+func (a *apiServer) ListDatum(ctx context.Context, request *pps.ListDatumRequest) (response *pps.DatumInfos, retErr error) {
+	func() { a.Log(request, nil, nil, 0) }()
+	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
+	jobInfo, err := a.InspectJob(ctx, &pps.InspectJobRequest{
+		Job: &pps.Job{
+			ID: request.Job.ID,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	pfsClient, err := a.getPFSClient()
+	if err != nil {
+		return nil, err
+	}
+	// Check 'stats' branch exists
+	branches, err := pfsClient.ListBranch(ctx, &pfs.ListBranchRequest{jobInfo.OutputCommit.Repo})
+	if err != nil {
+		return nil, err
+	}
+	var statsBranch *pfs.BranchInfo
+	for _, branch := range branches.BranchInfo {
+		if branch.Name == "stats" {
+			statsBranch = branch
+		}
+	}
+	if statsBranch == nil {
+		return nil, fmt.Errorf("stats branch not found on %v", jobInfo.OutputCommit.Repo.Name)
+	}
+	// List the files under /jobID to get all the datums
+	file := &pfs.File{
+		Commit: statsBranch.Head,
+		Path:   fmt.Sprintf("/%v", request.Job.ID),
+	}
+	allFileInfos, err := pfsClient.ListFile(ctx, &pfs.ListFileRequest{file})
+	if err != nil {
+		return nil, err
+	}
+	datums := make(map[string]*pps.DatumInfo)
+	for _, fileInfo := range allFileInfos.FileInfo {
+		fileInfo := fileInfo
+		_, datumHash := filepath.Split(fileInfo.File.Path)
+		datums[datumHash] = &pps.DatumInfo{
+			Datum: &pps.Datum{datumHash},
+			State: pps.DatumState_SKIPPED,
+		}
+	}
+
+	// Diff the files under /parentJobID and /jobID to get non-skipped datums
+	newFile := &pfs.File{
+		Commit: statsBranch.Head,
+		Path:   fmt.Sprintf("/%v", request.Job.ID),
+	}
+	commitInfo, err := pfsClient.InspectCommit(
+		ctx,
+		&pfs.InspectCommitRequest{
+			Commit: &pfs.Commit{
+				Repo: jobInfo.OutputCommit.Repo,
+				ID:   statsBranch.Head.ID,
+			},
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	oldFile := &pfs.File{
+		Commit: commitInfo.ParentCommit,
+		Path:   fmt.Sprintf("/%v", request.Job.ID),
+	}
+	resp, err := pfsClient.DiffFile(ctx, &pfs.DiffFileRequest{newFile, oldFile, true})
+	if err != nil {
+		return nil, err
+	}
+
+	// The newFileInfos contain datums that were new in this job
+	// For these datums, populate the status and stats
+	blacklist := map[string]bool{
+		"stats": true,
+		"logs":  true,
+		"pfs":   true,
+	}
+	pathToDatumHash := func(path string) (string, error) {
+		_, datumHash := filepath.Split(path)
+		if _, ok := blacklist[datumHash]; ok {
+			return "", fmt.Errorf("value %v is not a datum hash", datumHash)
+		}
+		return datumHash, nil
+	}
+	var egGetDatums errgroup.Group
+	getDatumsSemaphore := make(chan struct{}, 200)
```

Comment:
We actually have a `Limiter` package now that simplifies this pattern a little bit.